src/protocol/boolean/equality_test.rs

```diff
@@ -0,0 +1,284 @@
+use crate::{
+    error::Error,
+    ff::PrimeField,
+    protocol::{
+        boolean::{multiply_all_shares, random_bits_generator::RandomBitsGenerator, RandomBits},
+        context::Context,
+        BasicProtocols, RecordId,
+    },
+    secret_sharing::Linear as LinearSecretSharing,
+};
+
+// Compare two arithmetic-shared values `a` and `b` to compute `[a == b]` without revealing `(a == b)` itself.
+//
+// This is an implementation of 6.4 Equality Test Protocol in "Multiparty Computation for Interval, Equality, and
+// Comparison Without Bit-Decomposition Protocol", Nishide & Ohta, PKC 2007.
+// <https://doi.org/10.1007/978-3-540-71677-8_23>
+//
+// Observation: `[a == b]` is equivalent to `[a - b == 0]`
+// So we compute `[d] = [a - b]` and focus on computing `[d == 0]`
+//
+// Strategy:
+//  1. Generate random r
+//  2. Reveal c = d + r
+//  3. Observe that if d == 0, then c == r
+//
+// Protocol:
+//  1. generate a bitwise and arithmetically shared random value r
+//  2. compute [c]_p = [d]_p + [r]_p
+//  3. Reveal c
+//  4. Compute whether all the bits of c are the same as [r]_B
+//
+/// # Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+#[allow(clippy::many_single_char_names)]
+pub async fn equality_test<F, C, S>(
+    ctx: C,
+    record_id: RecordId,
+    rbg: &RandomBitsGenerator<F, S, C>,
+    a: &S,
+    b: &S,
+) -> Result<S, Error>
+where
+    F: PrimeField,
+    C: Context + RandomBits<F, Share = S>,
+    S: LinearSecretSharing<F> + BasicProtocols<C, F>,
+{
+    use EqualityTestStep as Step;
+
+    let d = a.clone() - b;
+
+    let r = rbg.generate().await?;
+
+    // Mask `d` with random `r` and reveal.
+    let c: u128 = (d + &r.b_p)
+        .reveal(ctx.narrow(&Step::Reveal), record_id)
+        .await?
+        .as_u128();
+
+    let one = S::share_known_value(&ctx, F::ONE);
+
+    let transformed_bits = r
+        .b_b
+        .iter()
+        .enumerate()
+        .map(|(i, r_bit)| {
+            let c_i = (c >> i) & 1;
+            if c_i == 1 {
+                r_bit.clone()
+            } else {
+                one.clone() - r_bit
+            }
+        })
+        .collect::<Vec<_>>();
```

Comment:
Discussed offline. Basically, multiply_all_shares accepts a maximum of 64 things or it will panic (due to BitOpStep). This function is not meant for long lists - so it's not worth it to do an iterator.